### ISSUE TITLE
twistd web options should be composable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3752]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jack reported|
|-|-|
|Trac ID|trac#3752|
|Type|enhancement|
|Created|2009-04-08 03:20:47Z|

Currently one can specify --path or --wsgi to twistd web to put either a static resource or a WSGI application at the root of the server.  It would be really great if these were composable along with a specification of where they should go.  Glyph suggested a syntax like:

```
twistd web --wsgi my.application --at /myapp --path /some/content --at /static
```

<details><summary>Searchable metadata</summary>

```
trac-id__3752 3752
type__enhancement enhancement
reporter__jack jack
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__tape tape
time__1239160847000000 1239160847000000
changetime__1239688997000000 1239688997000000
version__None None
owner__ 
cc__jack cc__exarkun cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jack commented|
|-|-|

[#3753](https://github.com/twisted/twisted/issues/3753) is the bug for the overlay implementation which would allow this kind of flexibility at the command line.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @glyph|
|-|-|

So, I think that depending on [#3753](https://github.com/twisted/twisted/issues/3753) for this is a bit overly ambitious.

The goal of this ticket, as I understand it, is taking `twistd web` from being something that's useful only for simple stuff, that can't walk and chew gum at the same time, to something that's functional enough to support more than one resource at a time.  Just enough to support practical use-cases where you have N things that you want your web server to be doing.

The archetypical site that I'd like `twistd web` to be able to support is one which has 2 or 3 site-wide applications, some static content, and maybe serves some stuff up out of home directories.

The "overlay resource" stuff is overkill for that goal, and probably counterproductive.  That (implicitly) supports arbitrary legacy applications, and various other difficult-to-specify stuff.

So, my scaled back proposal for this ticket is to refactor the command-line options to twisted.web so that it's clear which options create a resource, and which options modify the behavior of a resource.  Consider that the nonsensical command-line "twistd web --wsgi foo --path bar" will currently succeed with no error message.  Then, group them such that you are allowed to specify _one_ segment of the path, which will then call `putChild("segment", yourResource)` on whatever the root resource is.

So, a new suggestion for the way the UI for this might be presented on the command-line:

```
#!sh
$ twistd web --path /some/content --add my-app --wsgi my.app
# /some/content served at /,
#   my.app served at /my-app
$ twistd web --add content --path /some/content --add my-app \
     --wsgi my.app --add "" welcome.html
# /some/content served at /content, 
#   my.app served at /my-app,
#   / renders as the contents of welcome.html
#   everything else a 404
$ twistd web --add foo/bar/baz --path /some/content
# error, no slashes allowed
$ twistd web --wsgi my.app --path /some/content
# error, conflict on root resource
$ twistd web --add stuff --wsgi my.app --add stuff --path /some/content
# error, conflict on "stuff"
```
Well-behaved legacy sites that live under a given URL segment and just about any legitimate use-case for a _new_ application can easily be satisfied with this simple approach.

As a separate ticket, these different resource-creation options can be refactored into a plugin mechanism, or a plugin mechanism can be added (whichever makes more sense).

Wackier legacy sites and deliberately obscure dispatch schemes can be implemented in terms of plugins which talk to that mechanism.  In the meanwhile, '`twistd web --resource-script=...` or `--class=...` should serve to allow the construction of arbitrarily complex URL hierarchies without getting into the complexities of either .tac files or `twistd` plugins.

Jack, does this sound good?

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|jack commented|
|-|-|

Just to summarize, you want to limit the commandline options to top level children only, but offer more complex arrangements via resource script or something similar?

That sounds fine to me.  The use cases for the complex paths are probably rarer.  I don't see a super compelling reason not to allow them, but I assume you do.

As for prefixing --add versus postfixing --at, I prefer the postfixed --at syntax.  The distinction between modifying and adding still seems to be there to me.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Replying to [jack](#note_3):
> Just to summarize, you want to limit the commandline options to top level children only, but offer more complex arrangements via resource script or something similar?

Yes.  Although perhaps a better way to put it would be that '--at' is one resource mapping mechanism (a simple one) and more complex ones could be spelled differently, later.

> That sounds fine to me.  The use cases for the complex paths are probably rarer.  I don't see a super compelling reason not to allow them, but I assume you do.

It's not so much that I don't want to allow them as that the mechanism to traverse them is highly controversial.  How do you specify a path?  Routes?  `mod_rewrite` regexes?  Chains of python expressions?  How do you handle potential conflicts?  The number of options for URL dispatch is staggering, and really we don't need a powerful mechanism at all here.  Starting with the simplest possible thing, it's always easy to add features later.  It's harder to take them out :).

> As for prefixing --add versus postfixing --at, I prefer the postfixed --at syntax.  The distinction between modifying and adding still seems to be there to me.

OK.  Prefix syntax makes the error handling a little easier, but only a little; at the end of the day it's all the same stuff.  If you like `--at`, we can do that.  (I hope that one day we'll have a GUI application to do this instead of command-line args anyway :-)).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
